src/sync/barrier.rs

```diff
@@ -0,0 +1,130 @@
+use crate::sync::Mutex;
+use broadcaster::BroadcastChannel;
+
+/// A barrier enables multiple threads to synchronize the beginning of some computation.
+///
+/// ```
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use std::sync::Arc;
+/// use async_std::sync::Barrier;
+/// use futures_util::future::join_all;
+///
+/// let mut handles = Vec::with_capacity(10);
+/// let barrier = Arc::new(Barrier::new(10));
+/// for _ in 0..10 {
+///     let c = barrier.clone();
+///     // The same messages will be printed together.
+///     // You will NOT see any interleaving.
+///     handles.push(async move {
+///         println!("before wait");
+///         let wr = c.wait().await;
+///         println!("after wait");
+///         wr
+///     });
+/// }
+/// // Will not resolve until all "before wait" messages have been printed
+/// let wrs = join_all(handles).await;
```

Comment:
We make this example even more similar to [stdlib's](https://doc.rust-lang.org/std/sync/struct.Barrier.html#examples) by doing:

```rust
for handle in handles {
    handle.await;
}
```

This allows us to not have to use the `join_all` macro. It works especially well if we remove the `wr` result part from the example as well.

Reply:
fixed